### ISSUE TITLE
MCP42050 Digital Potentiometer issue, library MCP_POT.h

### ISSUE THREAD

**Issue by OvidiuBodo** [RobTillaart/MCP_POT#4]
Hello! As I said in the email, I'm back here with my problem.

I use an MCP42050 digital potentiometer to control the amplitude (gain) of a signal. I use Arduino Nano and your library. I am attaching my code so you can understand what it is about.

The hardware connections made by me are:
D13- Clock MCP42050, D11- MOSI, D12- MISO, PW1- I connected it to the inverting terminal of the amplifier, PA1- ground, PB1- here I apply my signal.

Can you help me identify the problem?

This is the code:
```cpp
#include "AD9833.h"
#include <SPI.h>
#include <Wire.h>
#include <Adafruit_GFX.h>
#include <Adafruit_SSD1306.h>
#define ENCODER_OPTIMIZE_INTERRUPTS
#include <Encoder.h>
#include <MCP_POT.h>

//dimensiunile afisajului OLED
#define SCREEN_WIDTH 128
#define SCREEN_HEIGHT 64

// denimin pinii pentru AD9833
#define FSYNC_PIN 10
#define SDATA_PIN 11
#define SCLK_PIN 13

// definim pinii afisajului OLED
#define OLED_MOSI   11
#define OLED_CLK    13
#define OLED_DC     9
#define OLED_CS     7
#define OLED_RESET  8

// pin buton (Frecventa->Amplitudine->Offset)
#define BUTTON_PIN  5
#define BUTTON_PIN2 4

// pini encoder ( pinul ENCODER_BTN este pinul switch-ului din encoder)
#define ENCODER_PIN_A 2
#define ENCODER_PIN_B 3
#define ENCODER_BTN 6

Adafruit_SSD1306 display(SCREEN_WIDTH, SCREEN_HEIGHT, &SPI, OLED_DC, OLED_RESET, OLED_CS);
Encoder myEncoder(ENCODER_PIN_A, ENCODER_PIN_B);


int buttonState = HIGH;
int lastButtonState = HIGH;
int encoderButtonState = HIGH;

int lastEncoderButtonState = HIGH;
int buttonPressCounter = 0;
bool initialMessageDisplayed = false;
long encoderPosition = 0;
long lastEncoderPosition = -999999999;
long incrementStep = 1;     // increment folosit la encoder
long currentValue = 0;     // valoarea care se afiseaza pe display

int initializare_loop = 0;
int val_amplitudine = 0;
long newEncoderPosition = myEncoder.read()>>1;

int indicator_stare_initiala = 0;
long currentValueFrecventa = 0;
long currentValueAmplitudine = 0;
long currentValueOffset = 0;
int contor_setare_parametru = 1;
int verificare =0;
//=========================
const int CS_PIN = A4;   // Pinul pentru Chip Select
const int MOSI_PIN = 11; // Pinul pentru Master Out Slave In
const int MISO_PIN = 12; // Pinul pentru Master In Slave Out
const int SCK_PIN = 13;  // Pinul pentru Clock

void displayMessage(String mesage); 
AD9833 gen(FSYNC_PIN);
MCP_POT potentiometer(CS_PIN, 0, 0, MOSI_PIN, SCK_PIN);
 

void setup(){

  pinMode(CS_PIN, OUTPUT);

  gen.Begin();
  Serial.begin(9600);
  
  SPI.begin();
  potentiometer.begin();

  display.begin(SSD1306_SWITCHCAPVCC);
  display.clearDisplay();
  pinMode(BUTTON_PIN, INPUT_PULLUP);
  pinMode(BUTTON_PIN2, INPUT_PULLUP);

  pinMode(ENCODER_BTN, INPUT_PULLUP);

  pinMode(ENCODER_PIN_A, INPUT);
  pinMode(ENCODER_PIN_B, INPUT);

  delay(10);

  displayMessage(0, 0, "Generator de semnale cu AD9833");
  delay(1500);
   //=================
  
}

//=========================================================================
void displayMessage(int x, int y, String message) {
  y = y * 10; // corectii pentru coordonate
  x = x * 6;
  display.setTextSize(1);
  display.setTextColor(SSD1306_WHITE, SSD1306_BLACK);
  display.setCursor(x, y);
  display.println(message);
  display.display();
}
//===================================================

void initializare(){

  while (initializare_loop == 0 ){
 
    if (digitalRead(BUTTON_PIN) ==1) {
      initializare_loop = 1;
      buttonPressCounter = 1;
    }
  }
}

void loop(){
  initializare();
 
  switch (buttonPressCounter) {
       
    case 1:{
      delay(400);
      myEncoder.write(0);
      display.clearDisplay();
      incrementStep =1;
      encoderPosition = 0;
      currentValueFrecventa = 0;
      currentValueAmplitudine = 0;
      currentValueOffset = 0;

      while (digitalRead(BUTTON_PIN) == 0){
        handleEncoder(&currentValueAmplitudine);
 
        displayMessage(0, 0, "Componenta continua:");
        displayMessage(1, 1, "Frecventa:"+String(currentValueFrecventa)+"  ");
        displayMessage(1, 2, "Amplitduine:"+String(currentValueAmplitudine));
        displayMessage(1, 3, "Offset:"+String(currentValueOffset));                            
      }
            
      buttonPressCounter = 2;
      break;
    }
           
    case 2:{
      delay(400);
      myEncoder.write(0);         
      display.clearDisplay();

      incrementStep =1;
      encoderPosition = 0;
      currentValueFrecventa = 0;
      currentValueAmplitudine = 0;
      currentValueOffset = 0;

      while (digitalRead(BUTTON_PIN) == 0){
        encoderButtonState = digitalRead(ENCODER_BTN);
        newEncoderPosition = myEncoder.read() >> 1; 
                      

        if(digitalRead(BUTTON_PIN2) == 0 ){
          contor_setare_parametru++;
          if(contor_setare_parametru > 3){
            contor_setare_parametru = 1;
          }
        }
            

        if (contor_setare_parametru == 1){
          handleEncoder(&currentValueFrecventa);
                  
        }         
        else if(contor_setare_parametru == 2){
          potentiometer.setValue(currentValueAmplitudine); // here I also tried with fixed values ​​such as 100, 50 but it did not work

          
          handleEncoder(&currentValueAmplitudine);
                  
        }
        else if(contor_setare_parametru == 3){
          handleEncoder(&currentValueOffset);      
        }

        displayMessage(0, 0, "Sinusoidal:      ");
        displayMessage(1, 1, "Frecventa:"+String(currentValueFrecventa)+"  ");
        displayMessage(1, 2, "Amplitduine:"+String(currentValueAmplitudine)+"  ");
        displayMessage(1, 3, "Offset:"+String(currentValueOffset)+"  ");

        gen.ApplySignal(SINE_WAVE, REG0, currentValueFrecventa);   
        gen.EnableOutput(true);
                  
      }
      buttonPressCounter = 3;
      break;
    }

    case 3:{
      delay(400);
      myEncoder.write(0);
          
      display.clearDisplay();

      incrementStep =1;
      encoderPosition = 0;
      currentValueFrecventa = 0;
      currentValueAmplitudine = 0;
      currentValueOffset = 0;

      while (digitalRead(BUTTON_PIN) == 0){
        encoderButtonState = digitalRead(ENCODER_BTN);
        newEncoderPosition = myEncoder.read() >> 1; 
            
        if(digitalRead(BUTTON_PIN2) == 0){
          contor_setare_parametru++;

          if(contor_setare_parametru > 3){
            contor_setare_parametru = 1;
          }
        }

        delay(40);
        if (contor_setare_parametru == 1){
          handleEncoder(&currentValueFrecventa);
        }           
        else if(contor_setare_parametru == 2){
          potentiometer.setValue(currentValueAmplitudine);
          handleEncoder(&currentValueAmplitudine);
        }           
        else if(contor_setare_parametru == 3){
          handleEncoder(&currentValueOffset);
        }

        displayMessage(0, 0, "Dreptunghiular:      ");
        displayMessage(1, 1, "Frecventa:"+String(currentValueFrecventa)+ "  ");
        displayMessage(1, 2, "Amplitduine:"+String(currentValueAmplitudine)+"  ");
        displayMessage(1, 3, "Offset:"+String(currentValueOffset)+"  ");

        gen.ApplySignal(SQUARE_WAVE, REG0, currentValueFrecventa);   
        gen.EnableOutput(true);
      }
      buttonPressCounter = 4;
      break;
    }

    case 4:{
      delay(400);
      myEncoder.write(0);
          
      display.clearDisplay();

      incrementStep =1;
      encoderPosition = 0;
      currentValueFrecventa = 0;
      currentValueAmplitudine = 0;
      currentValueOffset = 0;

      while (digitalRead(BUTTON_PIN) == 0){
        encoderButtonState = digitalRead(ENCODER_BTN);
        newEncoderPosition = myEncoder.read() >> 1; 

        if (digitalRead(BUTTON_PIN2) == 1 ){
          verificare = 0;
        }

        if(digitalRead(BUTTON_PIN2) == 0 && verificare == 0){
          contor_setare_parametru++;

          if(contor_setare_parametru > 3){
            contor_setare_parametru = 1;
          }
        }

        delay(40);
        if (contor_setare_parametru == 1){
          handleEncoder(&currentValueFrecventa);
        }
        else if(contor_setare_parametru == 2){
          handleEncoder(&currentValueAmplitudine);
        }
        else if(contor_setare_parametru == 3){
          handleEncoder(&currentValueOffset);
        }

        displayMessage(0, 0, "Triunghiular:      ");
        displayMessage(1, 1, "Frecventa:"+String(currentValueFrecventa)+"  ");
        displayMessage(1, 2, "Amplitduine:"+String(currentValueAmplitudine)+"  ");
        displayMessage(1, 3, "Offset:"+String(currentValueOffset)+"  ");

        gen.ApplySignal(TRIANGLE_WAVE, REG0, currentValueFrecventa);   
        gen.EnableOutput(true);          
      }
      buttonPressCounter = 1;
      break;
    }
  }
}
   
//=====================================================================
void handleEncoder(long *currentValue) {//------
  newEncoderPosition = myEncoder.read() >> 1; 

  if (digitalRead(ENCODER_BTN) == 1 ){
    indicator_stare_initiala = 0;
  }

  if (digitalRead(ENCODER_BTN) == 0 && indicator_stare_initiala == 0  ){//--
    incrementStep *= 10;
    if (incrementStep > 1000000) incrementStep = 1;
    indicator_stare_initiala =1;
  }

  if (encoderPosition != newEncoderPosition) {//---
    *currentValue = *currentValue + ((newEncoderPosition - encoderPosition) * incrementStep);
    encoderPosition = newEncoderPosition; // Actualizați poziția curentă a encoderului
  }
}
```

**Comment by RobTillaart** [comment]
> Can you help me identify the problem?

You did not tell what the problem is, so guessing the cause is rather difficult.

it is another question than you asked for in the mail:

> Can you help me with a code that has a function where I can enter values between 0-255, so that it changes the resistance value of the potentiometer?

OK, if you run the demo program that comes with the library, does the pot work?

What should the program do, what do you expect?

----

TIP:  make separate functions per case, as this makes the code more readable,

----

So some guesses.

**MCP_POT potentiometer(CS_PIN, 0, 0, MOSI_PIN, SCK_PIN);**
Think the two 0 's should be 255 to be sure you not interfere with the Serial IO.

**currentValueAmplitudine = 0;**
Maybe you reset this one too often?

**Comment by OvidiuBodo** [comment]
Application description: I have an AD9833 signal generator that generates 3 waveforms (sine, square and triangle). The frequency of the signal is controlled by means of an encoder (currentValueFrecventa).
This is an input parameter in the generator function.
I want to achieve a variable amplification of the signal, so I used an operational amplifier in inverting connection to do this.
I want to use the digital potentiometer as an input resistor, so that I can change the ratio R2/R1 (Feedback resistor/ non-inverting input resistor), thus creating a signal with variable amplitude. The value given as an input parameter for the digital potentiometer (the one between 0 and 255) I want to be currentValueAmplitudine.

**Comment by RobTillaart** [comment]
OK understood.

Have you tried to control the potentiometer only in a minimal sketch?
That would confirm the hardware works, that is the first level to diagnose.

**Comment by OvidiuBodo** [comment]
Yes. It works.

**Comment by RobTillaart** [comment]
What do you do differently in your application, compared to the minimal sketch?

**Comment by OvidiuBodo** [comment]
In the minimal application I controlled the light intensity through a led. But I used another code that I no longer have.

**Comment by RobTillaart** [comment]
What does your application not do?

**Comment by OvidiuBodo** [comment]
The potentiometer does not change its resistance in such a way as to change the ratio R2/R1. So I can't control the gain.

**Comment by RobTillaart** [comment]
As said before, you reset the amplitude too often.
You only need to do that in setup()

comment all lines that look like
```cpp
currentValueAmplitudine = 0;
```
and try again

**Comment by RobTillaart** [comment]
See
```cpp
    case 3:{
      delay(400);
      myEncoder.write(0);
          
      display.clearDisplay();

      incrementStep =1;
      encoderPosition = 0;
      currentValueFrecventa = 0;
      currentValueAmplitudine = 0;   <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<  set to zero
      currentValueOffset = 0;

      while (digitalRead(BUTTON_PIN) == 0){
        encoderButtonState = digitalRead(ENCODER_BTN);
        newEncoderPosition = myEncoder.read() >> 1; 
            
        if(digitalRead(BUTTON_PIN2) == 0){
          contor_setare_parametru++;

          if(contor_setare_parametru > 3){
            contor_setare_parametru = 1;
          }
        }

        delay(40);
        if (contor_setare_parametru == 1){
          handleEncoder(&currentValueFrecventa);
        }           
        else if(contor_setare_parametru == 2){
          potentiometer.setValue(currentValueAmplitudine);   <<<<<<<<<<<<<<<<  therfor set to zero !!!
          handleEncoder(&currentValueAmplitudine);            <<<<<<<<<<  get a new value afterwards ...
        }           
        else if(contor_setare_parametru == 3){
          handleEncoder(&currentValueOffset);
        }

        displayMessage(0, 0, "Dreptunghiular:      ");
        displayMessage(1, 1, "Frecventa:"+String(currentValueFrecventa)+ "  ");
        displayMessage(1, 2, "Amplitduine:"+String(currentValueAmplitudine)+"  ");  <<<<<<<<<  what value is printed here?
        displayMessage(1, 3, "Offset:"+String(currentValueOffset)+"  ");

        gen.ApplySignal(SQUARE_WAVE, REG0, currentValueFrecventa);   
        gen.EnableOutput(true);
      }
      buttonPressCounter = 4;
      break;
    }
```

**Comment by OvidiuBodo** [comment]
displayMessage(1, 2, "Amplitduine:"+String(currentValueAmplitudine)+"  ");  <<<<<<<<<  what value is printed here?

The presented value is the current value, i.e. the count made with the encoder.

**Comment by OvidiuBodo** [comment]
> As said before, you reset the amplitude too often. You only need to do that in setup()
> 
> comment all lines that look like
> 
> ```c++
> currentValueAmplitudine = 0;
> ```
> 
> and try again

ok, but when we are in a case, only he is executed. Does what is declared in the other cases influence him?

**Comment by RobTillaart** [comment]
> displayMessage(1, 2, "Amplitduine:"+String(currentValueAmplitudine)+" "); <<<<<<<<< what value is printed here?
> 
> The presented value is the current value, i.e. the count made with the encoder.

I understand, but what is the value is printed on the display?

**Comment by RobTillaart** [comment]
> > As said before, you reset the amplitude too often. You only need to do that in setup()
> > comment all lines that look like
> > ```c++
> > currentValueAmplitudine = 0;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and try again
> 
> ok, but when we are in a case, only he is executed. Does what is declared in the other cases influence him?

Yes as there is only one variable with that name.
and it is reset in case 3 and it is send to the potentiometer in case 3 (as shown in posted code above)/

**Comment by OvidiuBodo** [comment]
numbers like: 0, 1,2,.....100000.
Shall I send you a photo?Maybe it helps you understand better.

**Comment by RobTillaart** [comment]
> numbers like: 0, 1,2,.....100000. Shall I send you a photo?Maybe it helps you understand better.

I can very well imagine the numbers, 
however please send a photo that its value is between 200 and 250 or so.

Note that if a number like 100000 is displayed and you send that to the potentiometer with setValue() it will be truncated to an byte between 0..255 so it will have strange effects...

**Comment by RobTillaart** [comment]
Can you add the line
```cpp
Serial.println(currentValueAmplitudine);
before every call to 
potentiometer.setValue(currentValueAmplitudine);
```

What is the output (the values) printed on the Serial monitor?

**Comment by OvidiuBodo** [comment]
> Can you add the line
> 
> ```c++
> Serial.println(currentValueAmplitudine);
> before every call to 
> potentiometer.setValue(currentValueAmplitudine);
> ```
> 
> What is the output (the values) printed on the Serial monitor?

![image](https://github.com/RobTillaart/MCP_POT/assets/166039244/55458e22-f3f1-4948-9c69-a56accd211cd)
The value set by me using the encoder appears.

**Comment by OvidiuBodo** [comment]
> > numbers like: 0, 1,2,.....100000. Shall I send you a photo?Maybe it helps you understand better.
> 
> I can very well imagine the numbers, however please send a photo that its value is between 200 and 250 or so.
> 
> Note that if a number like 100000 is displayed and you send that to the potentiometer with setValue() it will be truncated to an byte between 0..255 so it will have strange effects...

![Uploading img2d.png…]()

**Comment by RobTillaart** [comment]
>  The value set by me using the encoder appears.

and what value appears on the serial monitor?   42 ? or 0? or ?

**Comment by OvidiuBodo** [comment]
> > The value set by me using the encoder appears.
> 
> and what value appears on the serial monitor? 42 ? or 0? or ?

72. This value changes between 0 and 255.

**Comment by OvidiuBodo** [comment]
Theoretically, the amplitude of my signal should have changed with the currentValueAmplitudine change, because the potentiometer changes its resistance.

**Comment by RobTillaart** [comment]
Can you measure the changing resistance when you change the amplitude parameter between 0..255?

It should change.

**Comment by OvidiuBodo** [comment]
I did a test.  I connected PA1 to GND.  PB1 to +5V.  PW1 output.  No matter what value I set (0-255), I have approximately 2V on the PW1 pin.

**Comment by RobTillaart** [comment]
That could indicate the device is broken.

**Comment by OvidiuBodo** [comment]
I will analyze again tomorrow.  As a CS pin, I used A4 because I no longer have digital pins available.  I don't think this affects performance negatively.  Regarding the software, is the function in which I give the value 0...255 as a parameter good?  I have an MCP41010 potentiometer and I would like to do some tests with it tomorrow.

**Comment by RobTillaart** [comment]
library is still labeled **experimental**, but it should work.
I'll go through datasheet again to check the handshake.

If it fails you could test if changing the SPI mode  (about line 150 in .cpp file) works: SPI_MODE1, SPI_MODE2, SPI_MODE3
MSBFIRST is definitely correct. 
```cpp
void MCP_POT::setSPIspeed(uint32_t speed)
{
  _SPIspeed = speed;
  _spi_settings = SPISettings(_SPIspeed, MSBFIRST, SPI_MODE0);
}
```

The CS is active LOW, also correct in the code

_3.5 Serial Clock (SCK)
This is the SPI port clock pin and is used to clock-in
new register data. Data is clocked into the SI pin on the
rising edge of the clock and out the SO pin on the falling
edge of the clock. This pin is gated to the CS pin (i.e.,
the device will not draw any more current if the SCK pin
is toggling when the CS pin is high). This pin has a
Schmitt Trigger input._

in the **swSPI_transfer()**, the data bit is set and then clock is made high so it should work.
```cpp
void  MCP_POT::swSPI_transfer(uint8_t val)
{
  uint8_t clk = _clock;
  uint8_t dao = _dataOut;
  //  MSBFIRST
  for (uint8_t mask = 0x80; mask; mask >>= 1)
  {
    digitalWrite(dao,(val & mask));
    digitalWrite(clk, HIGH);
    digitalWrite(clk, LOW);
  }
}
```

----

Did you fix the zeros in the constructor already?
(as that is from your Serial IO normally)

**Comment by RobTillaart** [comment]
```MCP_POT potentiometer(CS_PIN, 0, 0, MOSI_PIN, SCK_PIN);```

Found a possible conflict in your code:

you reuse the MOSI and SCK pins for the SW SPI of the potentiometer,
and these pins are also use for the HW SPI of the AD9833.

This may cause the two handshakes to interfere.

**Comment by OvidiuBodo** [comment]
I think the library had a problem or was creating a conflict somewhere. I used a different code structure and it works fine.

**Comment by RobTillaart** [comment]
Yes can you post how you solved it?

**Comment by OvidiuBodo** [comment]
Yes. 

I used this code found on the internet.

```cpp
#include <SPI.h>

const byte csPin           = A4;       // MCP42100 chip select pin
const int  maxPositions    = 256;     // wiper can move from 0 to 255 = 256 positions
const long rAB             = 100000;  // 100k pot resistance between terminals A and B
const byte rWiper          = 125;     // 125 ohms pot wiper resistance
const byte pot0            = 0x11;    // pot0 addr
const byte pot1            = 0x12;    // pot1 addr
const byte potBoth         = 0x13;    // pot0 and pot1 simultaneous addr
const byte pot0Shutdown    = 0x21;    // pot0 shutdown
const byte pot1Shutdown    = 0x22;    // pot1 shutdown
const byte potBothShutdown = 0x23;    // pot0 and pot1 simultaneous shutdown

void setup() {

  Serial.begin(9600);

  digitalWrite(csPin, HIGH);        // chip select default to de-selected
  pinMode(csPin, OUTPUT);           // configure chip select as output
  SPI.begin();

  Serial.println("MCP42100 Demo");
  Serial.println("First pot1 wiper moves from min to max in approx. 25% increments.");
  Serial.println("Then pot0 wiper moves from min to max and max to min in single increments");
  Serial.println();

}

void loop() {

  // move pot1 wiper manually to specific positions
  Serial.println("Pot 1 Control ");

  setPotWiper(pot1, 250);                    // minimum resistance
  delay(3000);

 
  // move pot0 wiper from min to max in increments of 1 position
 // Serial.println("Pot 0 Control ");
  for (int i = 0; i < 256; i++) {
    setPotWiper(pot0, i);
    delay(1000);
  }

  // move pot0 wiper from max to min in decrements of 1 position
 // for (int i = 256; i > 0; i--) {
   // setPotWiper(pot0, i);
   // delay(1000);
 // }

}

void setPotWiper(int addr, int pos) {

  pos = constrain(pos, 0, 255);            // limit wiper setting to range of 0 to 255

  digitalWrite(csPin, LOW);                // select chip
  SPI.transfer(addr);                      // configure target pot with wiper position
  SPI.transfer(pos);
  digitalWrite(csPin, HIGH);               // de-select chip

  // print pot resistance between wiper and B terminal
 // long resistanceWB = ((rAB * pos) / maxPositions ) + rWiper;
 // Serial.print("Wiper position: ");
 // Serial.print(pos);
 // Serial.print(" Resistance wiper to B terminal: ");
 // Serial.print(resistanceWB);
 // Serial.println(" ohms");

}
```

I measured the voltage on PW1, depending on the value between 0-255 that I was giving. The tension changes.

**Comment by RobTillaart** [comment]
Thanks, I will study it coming week.
See if and where are the differences.

**Comment by RobTillaart** [comment]
> The tension changes. 

I assume you mean current in (milli)amperes.

Added code tags for syntax highlighting for readability.

**Comment by RobTillaart** [comment]
```cpp
void setPotWiper(int addr, int pos) {

  pos = constrain(pos, 0, 255);            // limit wiper setting to range of 0 to 255

  digitalWrite(csPin, LOW);                // select chip
  SPI.transfer(addr);                      // configure target pot with wiper position
  SPI.transfer(pos);
  digitalWrite(csPin, HIGH);               // de-select chip
 ...
}
```

Code looks similar (Hardware SPI), it uses three addresses where I use an "offset" to select pm 0 or pm 1 or both. 
The difference is that in the library uses **SPI.beginTransaction(settings)** and **SPI.endTransaction()**.
These transactions are explicit to be able to use devices that use different SPI-settings.

----

In your first post you use the software SPI
```cpp
MCP_POT potentiometer(CS_PIN, 0, 0, MOSI_PIN, SCK_PIN);
```

Can you replace it with the hardware SPI interface, to see if that works?
```cpp
MCP_POT potentiometer(CS_PIN, 255, 255, &SPI);
```

I'm also interested if your original sketch works if you include your found code instead of the library.

**Comment by OvidiuBodo** [comment]
Hello Rob!
I replaced software  SPI with hardware SPI  but doesn't worked.
With the code found, it works.

**Comment by RobTillaart** [comment]
I ordered hardware (2x MCP42010) for testing.

**Comment by RobTillaart** [comment]
Just to let you know, the order has still not arrived...

**Comment by OvidiuBodo** [comment]
There is no problem. I used the solution I found in my application.
I am waiting for a feedback from you. I am curious if it will work in the tests you will do.

**Comment by RobTillaart** [comment]
Got the digi-pot MCP42010 yesterday evening

Wrote a test sketch and tested both SW SPI and HW SPI with an UNO a few minutes ago.
Worked well on the UNO.

The test sketch will be added to the examples in a next release. 
Going to test with different HW SPI speeds to see the working range.

**Comment by RobTillaart** [comment]
UNO works at 8 MHz SPI.

~~Found first bug - setting 2 channels at once  fails.~~ 
(my error)

**Comment by RobTillaart** [comment]
The default SPI speed was set to 8 MHz which is outside the datasheet range.
In the upcoming release this will be set back to 1 MHz.

Need to check how to handle SW SPI max speed as an ESP32 can be pretty fast, in fact on par with HW SPI.

**Comment by RobTillaart** [comment]
@OvidiuBodo 
Created a develop branch for the upcoming 0.2.1 release.
It might be that the reduction of the HW SPI clock solved your problem.
If you have time, could you give it a try?

**Comment by OvidiuBodo** [comment]
Hello! I will try during the next week.

_Bodo Ovidiu_
_Specializarea:Electronica aplicata_
_Grupa:332_
_An:III_
_Tel:0751466558_ 
 
  Pe joi, apr. 25, 2024 la 14:33, Rob ***@***.***> a scris:   


@OvidiuBodo
Created a develop branch for the upcoming 0.2.1 release.
It might be that the reduction of the HW SPI clock solved your problem.
If you have time, could you give it a try?

**Comment by RobTillaart** [comment]
@OvidiuBodo 
FYI, released the new 0.2.1 version.

**Comment by RobTillaart** [comment]
As the original problem is solved I close the issue.
